core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -500,6 +500,25 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
       _ <- (l.await *> ((self provide r) to p)).fork
     } yield l.succeed(()) *> p.await
 
+  /**
+   * Returns an effect that, if evaluated, will return the lazily computed result
+   * of this effect. The result of this effect will be recomputed with a fixed
+   * duration `d`. Also returns a canceler, which can be used to cancel the
+   * recomputation.
+   */
+  final def memoizeTimed(d: Duration): URIO[R with Clock, (IO[E, A], Canceler)] = {
```

Comment:
The `Canceler` makes the API trickier to use.

I'd suggest an alternate approach:

1. Go with `Ref[Option[Promise[...]]]` as the core backing type.
2. When requested, fill the ref with a new promise that will be completed with the result of the effect.
3. After the promise is completed, set a timer, and after the timer elapses, the ref will be cleared to None.
4. Every get will trry to get the completion of the promise if it exists, but if it doesn't exist, it will go to (2).

This implementation will not require a canceler.

I would also call this method `cached` and not `memoizeTimed`.